### PR TITLE
Feature/Termination Function

### LINKED ISSUE #760: [FEATURE] Signal when experiment has stopped
### Please describe the purpose of the feature. Is it related to a problem?
When experiments in Mava have finished, they are currently terminated using launchpad's `lp.stop` method which kills the entire process and no information is returned. 

### Describe the solution you'd like
A great feature would be if a system could return a signal when it has completed and not kill the entire process. This way multiple experiments could be run sequentially so that benchmarking could be done over an entire suite of environment, task and random seed configurations. 

### How do we know when implementation of this feature is complete?
Checklist:
- Systems can terminate while not stopping the entire process 
- Systems can return a flag once a given experiment has completed

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
Add a function to terminate the proceeses without killing the runtime
## Why?
When experiments in Mava have finished, they are currently terminated using launchpad's lp.stop method which kills the entire process and no information is returned.
## How?
Add `termination_fn` in `mava.utils.lp_utils` that will be called in `Terminator` component
## Extra
Close #760 
To test the function, you need to add a termination condition in the system (e.g. `termination_condition={"executor_steps": 5000},`)

### CODE REVIEW

File: mava/systems/jax/launcher.py [hunk 1]
```diff
@@ -117,6 +118,9 @@ def add(
 
         if self._multi_process:
             with self._program.group(name):
+                # Save the current PID to manage the termination of the process
+                node_fn.__self__.store.manager_pid = os.getpid()
```
[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/jax/launcher.py:122]: I believe we can't set it in the terminator component attributes because we need the pid of the launcher, I will be thankful if you have another option to deal with this pid and get it as an input to the terminator

[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/jax/launcher.py:123]: I will check if they have same pid if yes we can do that in the builder side

[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/jax/launcher.py:123]: Yes it is the same I will change it to be in `on_building_init_end`

[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/jax/launcher.py]: Done!

File: mava/components/jax/building/distributor.py [hunk 2]
```diff
@@ -144,7 +144,10 @@ def on_building_launch(self, builder: SystemBuilder) -> None:
         Returns:
             None.
         """
-        builder.store.program.launch()
+        worker_manager = builder.store.program.launch()
+
+        if worker_manager:
+            builder.store.worker_manager = worker_manager
```
Comment (arnupretorius) [inline mava/components/jax/building/distributor.py:150]: Slight suggested tweak: have `builder.store.worker_manager = builder.store.program.launch()` and then inside system.py's `launch` just return `self._builder.store.worker_manager` instead of having the if statement for the returns.

File: mava/systems/jax/system.py [hunk 3]
```diff
@@ -183,3 +185,8 @@ def launch(self) -> None:
 
         # Launch system
         self._builder.launch()
+
+        if self._builder.store.worker_manager:
```
Comment (arnupretorius) [inline mava/systems/jax/system.py:189]: See related comment above.

File: mava/utils/lp_utils.py [hunk 4]
```diff
@@ -86,6 +88,24 @@ def partial_kwargs(function: Callable[..., Any], **kwargs: Any) -> Callable[...,
...
+    parameter_server: SystemParameterServer,
+) -> None:
+    """Terminate the process
+
+    Args:
+        parameter_server: SystemParameterServer in order to get main pid
+    """
+    if parameter_server.store.manager_pid:
+        # parent_pid: the pid of the main thread process
+        parent_pid = parameter_server.store.manager_pid
```
Comment (arnupretorius) [inline mava/utils/lp_utils.py:101]: I can't seem to find where `parameter_server.store.manager_pid` is set in this PR? Was it added in another PR?

**Review by arnupretorius — COMMENTED** [review]
Thanks @OmaymaMahjoub, this is super useful! 🙂 

Just see my few comments.

File: mava/systems/jax/ippo/components.py [hunk 5]
```diff
@@ -67,6 +68,7 @@ def on_building_init_end(self, builder: SystemBuilder) -> None:
             None.
 
         """
+        builder.store.manager_pid = os.getpid()
```
[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/jax/ippo/components.py:71]: Yeah it is not related to IPPO only I will move it in `on_building_init_start` since it is a common predefined method!

File: examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        sample_batch_size=5,
+        num_epochs=config["num_epochs"],
+        num_executors=1,
+        multi_process=True,
+        clip_value=False,
+        termination_condition={"executor_steps": 10000},
+    )
+    # Launch the system.
+    worker_manager = system.launch()
+    # Wait for worker to finish - this critical to waiting for program to finish.
```
[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply examples/jax/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:114]: In the case of single process we don't have worker_manager since we don't use launchpad and we don't need the termination function so probably I will need to recheck that we return None or we don't return anything in the case of single process and that it works properly.

File: mava/utils/lp_utils.py (hunk 4 above)
Comment (OmaymaMahjoub) [reply mava/utils/lp_utils.py:101]: In the `launcher.py` the parameter_server will get a copy from the builder store and within this store it will receive the manager_pid that was initialized in `on_building_init_end`

Comment (arnupretorius) [reply mava/utils/lp_utils.py:101]: Ah yes, ok makes sense, thanks.

File: mava/systems/jax/ippo/components.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arnupretorius) [reply mava/systems/jax/ippo/components.py:71]: It can stay in the hook `on_building_init_end`, but just thought we might want to include it in a more general component, than keeping it inside a component defined just for ippo. For example, instead of putting it in `ExtrasLogProbSpec` component, we add it to a more general component that will be used by other systems as well, e.g. the `Distributor` component. Or something like that, where it makes sense.

[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/jax/ippo/components.py:71]: That make sense, okay I will thanks!

[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply mava/systems/jax/ippo/components.py]: Done now it is in the `mava/components/building/environments.py` on_building_init_start method

File: mava/components/jax/building/distributor.py (hunk 2 above)
Comment (OmaymaMahjoub) [reply mava/components/jax/building/distributor.py]: I think it will be a bit tough to do that since we can't easily access the builder from the launch function since we use the function this way `builder.store.program.launch() ` I tried to find a solution and I think it is possible but it will make the code more complex.
If you think this is necessary then I will be happy to change it!

File: examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    """Runs lp inside a new thread.
+
+    Args:
+        _ : unused.
+    """
+    config = {}
+    # Loop through a hyperparam such as num_epochs
+    for num_epochs in [5, 10, 15]:
+        config["num_epochs"] = num_epochs
+        create_and_run_lp_program(config)
```
[the remark this answers is not in the document]
Comment (OmaymaMahjoub) [reply examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:129]: Probably @KaleabTessera can answer you about that

File: examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        num_epochs=config["num_epochs"],
+        num_executors=1,
+        multi_process=True,
+        clip_value=False,
+        termination_condition={"executor_steps": 10000},
+    )
+    # Launch the system.
+    worker_manager = system.launch()
+    # Wait for worker to finish - this critical to waiting for program to finish.
+    if worker_manager:
```
Comment (DriesSmit) [inline examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:115]: Can we maybe move this inside the `system.launch`? Instead of needing to put it in every example? We can create a function `system.launch(wait=True)`.

**Review by DriesSmit — CHANGES_REQUESTED** [review]
Thanks for the updates @OmaymaMahjoub :slightly_smiling_face:

File: examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (KaleabTessera) [reply examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:129]: @DriesSmit Outside of tensorboad this isn't an issue since we log the config for each run. In tf, I think we should log hyperparams/config as opposed to changing the id since we could have multiple configs params that we sweep. I will create an issue for this.

File: examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py (hunk 8 above)
Comment (KaleabTessera) [reply examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:115]: @DriesSmit I get what you are saying. I think returning the worker manager might give us more flexibility in the future, maybe we can do things like pause exps or pbt. @OmaymaMahjoub What do you think?

**Review by KaleabTessera — APPROVED** [review]
Thanks @OmaymaMahjoub ! This is a excellent PR! All benchmarks for this look good!

File: examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py (hunk 8 above)
Comment (OmaymaMahjoub) [reply examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:115]: I agree with Kale-ab but we can make it an option if you want to!

Comment (OmaymaMahjoub) [reply examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:115]: Will work on it in another PR

File: examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (DriesSmit) [reply examples/debugging/simple_spread/feedforward/decentralised/run_ippo_grid_search.py:129]: Sounds good :slightly_smiling_face: Are we logging the the mava_id for each run as well? Then I'm happy :+1:

**Review by DriesSmit — APPROVED** [review]

**Review by arnupretorius — APPROVED** [review]
Thanks @OmaymaMahjoub 🤗

**Review by KaleabTessera — APPROVED** [review]
Thanks @OmaymaMahjoub ! :fire:
